### DISCUSSION TITLE
Bisq2 > user messaging

### DISCUSSION THREAD

**Discussion by skaunov** [bisq-network/bisq2#1146] [Q&A]
Hi, could you layout current designs/ideas if any on how messaging user will be upgraded compared to Bisq? Is there now anything beyond current Android notification app? Will trade and mediation messaging be touched?

**Comment by HenrikJannsen** [comment]
You mean the mobile notifications with the Bisq mobile app?
No there are currently no plans for supporting that, but would be nice at some point. But too much open work for the current devs...

**Comment by skaunov** [comment]
Actually I wanted to discuss if it's possible to unify messaging/notification channels with modern E2EE. Thanks for sharing the plans. 

I can see how Matrix can fit and what challenges are there to employ it. (That would need to keep own system in place and use Matrix only for notification, i.e. only send out events, but not accept any info. Or decentralization will be damaged.)

Also I wanted to propose contact Briar maintainer and contributors, to check if they would be interested to integrate it into Bisq2. Seems like alignment of the projects approaches is high, so it could be main messaging system across the whole app, and attract users of both projects to the second project.

**Comment by HenrikJannsen** [comment]
Briar is a great project and very close ideas as our network layer. 

The problem with mobile (specially Apple) is that Tor based (or additional I2P for Bisq 2) is problematic. On Android its technically less problematic but from usability there might be issues, like for a notification app you don't want to wait 30-60 sec. until the Tor network is ready. Keeping it running in the background drains resources.

Also for notifications push notifications are essential, and again for Apple its harder to align with high privacy and decentralization requirements as with Android but both add some constraints.

To apply the applications network/messaging infrastructure to mobile is only partially possible. For mobile there will be different sets of trade-offs and we plan in future once someone works on an Bisq Easy mobile app, to offer different options to let the user decide if they prefer more convenience or more privacy/ censorship resistance [1].


If you have any more concrete ideas or suggestions please share.


Unfortunately there is no doc/overview about the network layer in Bisq 2. It is on our TODO list to create such a doc, but too much work for the small dev team...

## Here is a quick high level overview of the Bisq 2 P2P network

We use multiple privacy networks in parallel (currently Tor, I2P). This removes the single point of failure in Bisq 1 when Tor has issues, as it happened multiple times in the past.
Every message has a proof of work attached for DoS protection (Tor added a similar concept recently). The difficulty is determined by the receivers node load (how much network load the node has) and the message type and size.

There is a pure messaging part and a data storage part.

All messages are signed and encrypted and sent via the supported networks to the peer (e.g. a tor hidden service).
If the peer is offline, the data is packed (encrypted) into a mailbox message and posted to the networks data storage layer. Once the peer goes online they receive the data in the network and remove the mailbox message from the data storage.
This concept is also used to enable messaging between Tor and I2P if both users only support 1 network. The data storage layer is used as a relay by using the mailbox message concept.
All data have a time to life. Once expired the data gets removed locally. Each node stores all data. There are not big data objects (images, videos,...). Data storage size is expected to be in the 5-20 MB range. Each data types is stored independently and can a policy can be applied (e.g. max size) to protect against spam flood attacks.

There are different types of data to be stored in the network:
- Authenticated data: Data like offers (usually public), signed by the creator and only the creator can remove the data. 
- Mailbox data: Sender signs and encrypt data, only receiver can decrypt and remove data.
- Authorized data: There is a key pair used where the public key is well known (authorized by the Bisq DAO for special roles, like Bisq1 bridge node, mediator, moderator,...). The data's signature gets verified against the well known pub key (either hard coded or distributed via the bridge node to the network).

When a user goes online they connect first to the well known seed nodes. From them they get the peer addresses in the network. Then they request data from a random set of peers. Data delivery is size limited and a set of hashes of already obtained data is used for a diff to only get missing data. This will be repeated process until all data is received. Data is streaming in as small chunks at the startup of the app (a bit Bittorrent like), in contrast to Bisq 1 where we request all data at once, which has become a problem on slow connection as data is getting larger (5-10 MB).


[1] https://github.com/bisq-network/bisq2/discussions/298

**Comment by skaunov** [comment]
Thanks for the overview! Just wanted to highlight what led me to the suggestion. Briar is actively doing desktop version. Also they use Java too. So for user-facing communications it would be cute since no additional apps would be needed as you can just manage your notifications/chats right in the Briar app.

Unfortunately no more detailed proposals from me as I really superficially understand both apps. So maybe Bisq2 will supersede Briar in the chatting across platforms. That would be cool too!

**Comment by HenrikJannsen** [comment]
Bisq does not try to become a messaging app. Those have different requirements and properties. That's also the reason why we did not use any existing one. Its a special purpose app with special needs and constraints. Limiting dependencies to other projects is one of the desired properties of Bisq.

**Reply by skaunov** [reply]
That's great to hear! Didn't even know such a move was discussed.

Maybe I just don't see whole picture! Anyway current trading/dispute messaging as well as notifications are really painful. =( I will be wondering to see Bisq2 solution! Cheers!
